Introduce cilium operator to manage per cluster tasks

tgraf opened the pull request:
This introduces an operator that will run with a single replica per cluster. The purpose of the operator is to:
 * Perform janitorial tasks on the kvstore
 * Synchronize k8s state to the kvstore as required for clustermesh operation
 * Help with upgrade preparations
 * Monitor the etcd cluster
 * Translate CNPs as required

This PR establishes the skeleton and adds the required services and endpoints synchronization to prepare service discovery in a clustermesh environment.

For now, the operator lives in the main repository and versioning is directly tied to the Cilium version. It is debatable to move the operator into a separate repository but it will require a versioning strategy and a compatibility matrix with enforcement.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6171)
<!-- Reviewable:end -->

eloycoto commented on daemon/daemon_main.go:
```diff
@@ -347,6 +343,20 @@ func checkMinRequirements() {
+	flags.String(option.ClusterName, defaults.ClusterName, "Name of the cluster")
+	viper.BindEnv(option.ClusterName, option.ClusterNameEnv)
+	flags.BoolP("debug", "D", false, "Enable debugging mode")
+	flags.String(option.K8sAPIServerName, "", "Kubernetes api address server (for https use --k8s-kubeconfig-path instead)")
```
This should be: 
```
	flags.StringVar(&option.Config.K8sAPIServer, option.K8sAPIServerName, "", "Kubernetes api address server (for https use --k8s-kubeconfig-path instead)")
	flags.StringVar(&option.Config.K8sKubeConfigPath, option.K8sKubeConfigPathName, "", "Absolute path of the kubernetes kubeconfig file")
```

tgraf committed:
store: Factor out event handling into Observer interface

Observing events has been part of the Key interface. This worked great for
existing uses but is not ideal when object ownership and event handling is
split. Decouple the interface to allow for the introduction of global services.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
service: Add service definition for kvstore

Add a ClusterService type to represent services in the kvstore.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Add Dockerfile

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Add synchronization of Kubernetes services into kvstore

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
Introduce skeleton for cilium-operator

The operator not performing any tasks yet besides connecting to the Kubernetes
apiserver and to the kvstore.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Add deployment YAML and RBAC rules

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on examples/kubernetes/1.10/cilium-crio.yaml:
```diff
@@ -329,6 +329,119 @@ spec:
+          - "--debug=$(CILIUM_DEBUG)"
+          - "--kvstore=etcd"
+          - "--kvstore-opt=etcd.config=/var/lib/etcd-config/etcd.config"
+        env:
```
Consider adding a liveliness probe as we want the container to fail fast and restart immediately if it's not working.

aanm commented on operator/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+ARG V
+RUN make -C operator LOCKDEBUG=$LOCKDEBUG V=$V
+
+FROM ubuntu:18.04
```
Would be amazing if it would work with: 
```
CGO_ENABLED=0 GOOS=linux go build -a -ldflags '-extldflags "-static"' make -C operator
....
FROM scratch
COPY --from=builder /go/src/github.com/cilium/cilium/operator/cilium-operator /usr/bin/cilium-operator
CMD ["/usr/bin/cilium-operator"]
```

tgraf replied:
A bit more work is needed to remove the bpf golang dependencies from pkg/endpoint and pkg/k8s. We are on the way there.
